tests/infer/test_valid_models.py

```diff
@@ -2344,6 +2344,58 @@ def guide():
     assert_ok(model, guide, elbo)
 
 
+@pytest.mark.stage("funsor")
+@pytest.mark.parametrize("num_particles", [1, 2])
+def test_collapse_diag_normal_plate_normal(num_particles):
+    T, d = 5, 3
+    data = torch.randn((T, d))
+
+    def model():
+        x = pyro.sample("x", dist.Normal(0., 1.)).unsqueeze(-1)
+        with poutine.collapse():
+            with pyro.plate("data", T, dim=-1):
+                expand_shape = (d,) if num_particles == 1 else (num_particles, 1, d)
+                y = pyro.sample("y", dist.Normal(x, 1.).expand(expand_shape).to_event(1))
+                pyro.sample("z", dist.Normal(y, 1.).to_event(1), obs=data)
```

Comment:
This fails due to 2 reasons:
+ `y.output` is Reals[d], which will raise error while infer_param_domain in funsor: `Output mismatch: Reals[2] vs Real`
+ to_event is not available for Funsor distributions.

Reply:
Hmm I wonder if there is always enough information available in the args of expanded distributions to automatically determine the event dim, so that we could make `.to_event()` a no-op on funsors. For example here we could deduce the event shape from `y.output`.

@eb8680 would it be possible to support this in `to_funsor()` and `to_data()`?

Reply:
> would it be possible to support this in to_funsor() and to_data()?

It should be possible, at least in principle, and if we want `collapse` to work seamlessly with models that use `.to_event` we'll need something like that.  I think it will require changing the way type inference works in `funsor.distribution.Distribution`, though.

The point of failure in Funsor is the `to_funsor` conversion of parameters in `funsor.distribution.DistributionMeta.__call__`:
```py
class DistributionMeta(FunsorMeta):
    def __call__(cls, *args, **kwargs):
        kwargs.update(zip(cls._ast_fields, args))
        value = kwargs.pop('value', 'value')

        # Failure occurs here -------v
        kwargs = OrderedDict(
            (k, to_funsor(kwargs[k], output=cls._infer_param_domain(k, getattr(kwargs[k], "shape", ()))))
            for k in cls._ast_fields if k != 'value')

        # this is also incorrect
        value = to_funsor(value, output=cls._infer_value_domain(**{k: v.output for k, v in kwargs.items()}))

        args = numbers_to_tensors(*(tuple(kwargs.values()) + (value,)))
        return super(DistributionMeta, cls).__call__(*args)
```
In this test case, `kwargs[k]` is `loc = Variable("y", Reals[2])`, and `Normal._infer_param_domain(...)` thinks the output should be `Real` instead of `Reals[2]`.

I think a general solution (at least when all parameters have the same broadcasted `.output`) is to compute unbroadcasted parameter and value shapes up front, then broadcast:
```py
# compute unbroadcasted domains
domains = {k: cls._infer_param_domain(k, getattr(kwargs[k], "shape")) for k in kwargs if k != "value"}
domains["value"] = cls._infer_value_domain(cls, **domains)

# broadcast individual param domains with Funsor inputs
# this avoids .expand-ing underlying parameter tensors
for k, v in kwargs.items():
    if isinstance(v, Funsor):
        domains[k] = Reals[broadcast_shapes(v.shape, domains[k].shape)[0]]  # assume all Reals for exposition

# broadcast value domain, which depends on param domains, with broadcasted param domains
domains["value"] = Reals[broadcast_shapes(domains["value"], *domains.values())[0]]
```

Now the previously incorrect `to_funsor` conversions reduce to
```py
kwargs = OrderedDict((k, to_funsor(kwargs[k], output=domains[k])) for k in kwargs)
value = kwargs["value"]
```

WDYT?

Reply:
Sounds reasonable to me.

Reply:
Ok, I put up a draft Funsor PR here: https://github.com/pyro-ppl/funsor/pull/402

Reply:
Thanks, Eli! testing now...

Reply:
I believe what's in that PR after my recent push is sufficient for this test case, although it's not ready to merge because of some edge cases in the distributions. Let me know if it's not working.